python/cucim/src/cucim/core/operations/color/stain_normalizer.py

```diff
@@ -0,0 +1,223 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+
+import cupy as cp
+
+
+class HEStainExtractor:
+    """Extract stain coefficients from an image.
+
+    Parameters
+    ----------
+        source_intensity: transmitted light intensity.
+            Defaults to 240.
+        alpha: percentiles to ignore for outliers, so to calculate min and max,
+            if only consider (alpha, 100-alpha) percentiles. Defaults to 1.
+        beta: absorbance threshold for transparent pixels.
+            Defaults to 0.15
+
+    Note:
+        Please refer to this paper for further information on the method:
+        http://wwwx.cs.unc.edu/~mn/sites/default/files/macenko2009.pdf
+    """
+
+    def __init__(
+        self,
+        source_intensity: float = 240,
+        alpha: float = 1,
+        beta: float = 0.15,
+    ) -> None:
+        self.source_intensity = source_intensity
+        self.alpha = alpha
+        self.beta = beta
+
+    def _stain_decomposition(self, image: cp.ndarray) -> cp.ndarray:
+        """Extract the matrix of stain coefficient fromthe image.
+
+        Parameters
+        ----------
+            image: RGB image to perform stain extraction on
+
+        Returns
+        -------
+            stain_coeff: stain attenuation coefficient matrix derive from the
+                image, where first column is H, second column is E, and
+                rows are RGB value
+
+        """
+        # calculate absorbance
+        image = image.astype(cp.float32, copy=False) + 1.0
+        absorbance = -cp.log(
```

Comment:
Marcenko et. al. uses `log10` rather than `log` in the publication, but existing open source implementations I found all use the natural log. I guess either is valid and it only effects the absolute scale of the `beta` threshold. Empirically, computation of `log` is better optimized than `log10`, so we should probably keep this as is.

Reply:
You are right! It only affect `beta` and as long as we are using the correct inverse (`exp` for `log` and `10**` for `log10`) both are fine.